### PR TITLE
Add LazyHTML.css_path/1

### PR DESCRIPTION
**Pull request by ftes**
Implement `LazyHTML.css_path/1` via NIF.
Return a document-relative CSS path for each selected element.

```elixir
~S|<main><span>first</span><span>second</span></main>|
|> LazyHTML.from_fragment()
|> LazyHTML.query("span")
|> LazyHTML.css_path()

    [
      "main:nth-child(1) > span:nth-child(1)",
      "main:nth-child(1) > span:nth-child(2)"
    ]
```

To build each path, walk from selected element towards document root.
At every level record element's qualified tag name and its 1-based `:nth-child` position.
The resulting path can be passed back to `query/2` to find the same element in the original document or fragment.

As with `tag/1` and `nth_child/1`, text and comment nodes in the selection are skipped.

### Use case

I'm working on [ftes/fluffy](https://github.com/ftes/fluffy). It applies Playwright-style locators and actions to Phoenix-rendered HTML through LazyHTML. It uses structural paths to join locator results back to a document index for ancestry, document order, form ownership, actionability, and render reconciliation. A batched API provides that bridge without repeated NIF calls or reliance on private Lexbor pointers.

### Performance

`css_path/1` is the batched composition of `parent_node/1` and `nth_child/1` via a NIF.
The same composition implemented in Elixir would require several NIF calls.

See benchark below.

### Prior art

Comparable node-to-path APIs exist in [lxml](https://lxml.de/apidoc/lxml.etree.html), [Nokogiri](https://nokogiri.org/rdoc/Nokogiri/XML/Node.html#method-i-css_path), [jsoup](https://jsoup.org/apidocs/org/jsoup/nodes/Element.html#cssSelector()), and [Html Agility Pack](https://documentation.help/HtmlAgilityPack/44e12d6f-4137-9f94-0430-b4d6b0231a01.htm).

## Benchmark

This PR vs. Elixir-based composition of `tag/1`, `nth_child/1` and `parent_node/1`.

Output:
```
Name                                    ips        average  deviation         median         99th %
css_path/1 (native, this PR)        1.33 K        0.75 ms     ±5.11%        0.75 ms        0.88 ms
composed public API (Elixir)       0.0424 K       23.56 ms     ±2.08%       23.42 ms       25.60 ms

Comparison:
css_path/1 (native, this PR)        1.33 K
composed public API (Elixir)       0.0424 K - 31.34x slower +22.81 ms
````

<details>

<summary>

`tmp/benchmark.exs`

</summary>

```elixir
# Save in checked out workspace as tmp/benchmark.exs
# Run via `elixir tmp/benchmark.exs`

Mix.install([
  {:benchee, "~> 1.4"},
  {:lazy_html, path: File.cwd!()}
])

defmodule CSSPathBenchmark do
  def public_api_path(lazy_html) do
    Enum.map(lazy_html, &public_api_path(&1, []))
  end

  defp public_api_path(node, segments) do
    [tag] = LazyHTML.tag(node)
    [position] = LazyHTML.nth_child(node)
    segments = ["#{tag}:nth-child(#{position})" | segments]
    parent = LazyHTML.parent_node(node)

    if Enum.empty?(parent) do
      Enum.join(segments, " > ")
    else
      public_api_path(parent, segments)
    end
  end
end

items =
  Enum.map_join(1..1_000, fn index ->
    ~s|<li><a href="/#{index}">Item #{index}</a></li>|
  end)

links =
  "<main><ul>#{items}</ul></main>"
  |> LazyHTML.from_document()
  |> LazyHTML.query("a")

native = fn -> LazyHTML.css_path(links) end
elixir = fn -> CSSPathBenchmark.public_api_path(links) end

if native.() != elixir.() do
  raise "implementations returned different paths"
end

Benchee.run(
  %{
    "css_path/1 (native, this PR)" => native,
    "composed public API (Elixir)" => elixir
  },
  time: 3,
  warmup: 0
)
```

</details>

### CODE REVIEW

**Comment by ftes** [comment]
Just an idea. I can work around this missing feature - just thought it might be interesting for others.
I'm not a C++ expert, without LLM assistance I wouldn't have created this PR. I've focused on the benchmark and tests to verify behaviour.

File: lib/lazy_html.ex [hunk 1]
```diff
@@ -399,6 +399,34 @@ defmodule LazyHTML do
...
+  @doc """
+  Returns a document-relative CSS path for each selected element in `lazy_html`.
+
+  To build each path, LazyHTML walks from the selected element towards the
+  document root. At every level it records the element's qualified tag name
+  and its 1-based `:nth-child` position. The resulting path can be passed back
+  to `query/2` to find the same element in the original document or fragment.
+
+  As with `tag/1` and `nth_child/1`, text and comment nodes in the selection
+  are skipped.
```
Comment (jonatanklosko) [inline lib/lazy_html.ex]: We don't need to describe the implementation, I think this is enough:

```suggestion
  @doc """
  Returns a CSS path for each root element in `lazy_html`.

  Note that if there are text or comment root nodes, they are ignored,
  and they have no corresponding path in the result.
```

File: README.md [hunk 2]
```diff
@@ -47,8 +47,17 @@ hyperlinks = LazyHTML.query(lazy_html, "a")
...
+LazyHTML.css_paths(hyperlinks)
+#=> [
+#=>   "div:nth-child(1) > a:nth-child(1)",
+#=>   "div:nth-child(1) > a:nth-child(2)"
+#=> ]
 ```
 
+`LazyHTML.css_paths/1` returns one document-relative, queryable CSS selector
+for each selected element, while keeping the operation batched in native code.
+
```
Comment (jonatanklosko) [inline README.md]:
````suggestion
```

````

File: c_src/lazy_html.cpp [hunk 3]
```diff
@@ -902,6 +904,124 @@ std::vector<int64_t> nth_child(ErlNifEnv *env, ExLazyHTML ex_lazy_html) {
...
+  for (auto node : ex_lazy_html.resource->nodes) {
+    if (node->type != LXB_DOM_NODE_TYPE_ELEMENT) {
+      continue;
+    }
+
+    auto segments = std::vector<std::string>();
+    auto current = node;
+
+    while (current != NULL && current->type == LXB_DOM_NODE_TYPE_ELEMENT &&
+           !(is_fragment && lxb_html_tree_node_is(current, LXB_TAG_HTML))) {
```
Comment (jonatanklosko) [inline c_src/lazy_html.cpp:982]: Let's update [parent_node](https://github.com/ftes/lazy_html/blob/ffeb8035b39083da5b076ee80ed336592213d917/c_src/lazy_html.cpp#L865-L866) to also use this version of the conditional for consistency.

File: c_src/lazy_html.cpp [hunk 4]
```diff
@@ -902,6 +904,124 @@ std::vector<int64_t> nth_child(ErlNifEnv *env, ExLazyHTML ex_lazy_html) {
...
+      throw std::runtime_error("tag name is not valid UTF-8");
+    }
+
+    bool leading_digit = index == 0 && is_ascii_digit(codepoint);
+    bool digit_after_leading_hyphen =
+        index == 1 && data[0] == '-' && is_ascii_digit(codepoint);
+    bool control = codepoint <= 0x1F || codepoint == 0x7F;
+
+    // Lexbor's selector parser currently requires non-ASCII identifier code
+    // points to be escaped, even though CSS allows them unescaped.
```
Comment (jonatanklosko) [inline c_src/lazy_html.cpp]:
```suggestion
    // Lexbor's selector parser currently requires some non-ASCII identifier
    // code points to be escaped, even though CSS allows them unescaped
    // (see lxb_css_syntax_state_non_ascii). Instead of special casing those
    // ranges, we simply escape all non-ASCII code points.
```

File: lib/lazy_html.ex [hunk 5]
```diff
@@ -399,6 +399,34 @@ defmodule LazyHTML do
...
+      ...> |> LazyHTML.query("span")
+      ...> |> LazyHTML.css_paths()
+      [
+        "main:nth-child(1) > span:nth-child(1)",
+        "main:nth-child(1) > span:nth-child(2)"
+      ]
+
+  """
+  @spec css_paths(t()) :: list(String.t())
+  def css_paths(%LazyHTML{} = lazy_html) do
```
Comment (jonatanklosko) [inline lib/lazy_html.ex]: Let's rename it to `css_path`. We use singular names like `tag` and `parent_node`, and they are applied for every node in the batch.

File: c_src/lazy_html.cpp [hunk 6]
```diff
@@ -902,6 +904,124 @@ std::vector<int64_t> nth_child(ErlNifEnv *env, ExLazyHTML ex_lazy_html) {
...
+               codepoint == '-' || codepoint == '_') {
+      if (index == 0 && codepoint == '-' && length == 1) {
+        identifier.append("\\-");
+      } else {
+        identifier.push_back(static_cast<char>(codepoint));
+      }
+    } else {
+      identifier.push_back('\\');
+      identifier.push_back(static_cast<char>(codepoint));
+    }
```
Comment (jonatanklosko) [inline c_src/lazy_html.cpp:963]: We should have one test that asserts a few of these edge cases.

**Review by jonatanklosko — COMMENTED** [review]
Hey, thanks for the PR! I dropped a few comments and then we can ship it.

File: c_src/lazy_html.cpp (hunk 6 above)
Comment (ftes) [reply c_src/lazy_html.cpp:963]: Added.

File: c_src/lazy_html.cpp (hunk 3 above)
Comment (ftes) [reply c_src/lazy_html.cpp:982]: Changed

**Comment by ftes** [comment]
Thanks for your feedback Jonatan. It should all be addressed now.

File: README.md [hunk 7]
```diff
@@ -47,8 +47,6 @@ hyperlinks = LazyHTML.query(lazy_html, "a")
 
 LazyHTML.attribute(hyperlinks, "href")
 #=> ["https://elixir-lang.org", "https://www.erlang.org"]
-```
-
 LazyHTML also provides several high-level conveniences:
```
Comment (jonatanklosko) [inline README.md]:
````suggestion
```

LazyHTML also provides several high-level conveniences:
````
